### PR TITLE
CORTX-29727:generate addb plugin as part of cortx-rgw-integration rpm

### PR DESCRIPTION
**Pull request by sachitanands**
Generate addb plugin binary as a part of rgw-integration rpm

` sh build.sh -build_no 12345 -version 1.0.0 -git_hash 1g0d3gt2 -addb`
Signed-off-by: Sachitanand Shelake <sachitanand.shelake@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: jenkins/build.sh [hunk 1]
```diff
@@ -18,18 +18,59 @@ set -e
...
+RGW_REPO="$BASE_DIR"/../cortx-rgw
 
 usage() {
-    echo """usage: $PROG [-v version] [-g git_version] [-b build_number]""" 1>&2;
+    echo """usage: $PROG [-v version] [-g git_version] [-b build_number][-a]
+    where,
+          -v <version_id>   Specify rpm version
+          -g <git_version>  Specify git version hash
+          -b <build_number> Specify build version
+          -a                Generates addb plugin as part of cortx-rgw-integration""" 1>&2;
```
Comment (rahul-tripathi-git) [inline jenkins/build.sh:35]: Ttripple quotes are not required here, (bash script). /n might be needed. Please check once.

File: jenkins/build.sh [hunk 2]
```diff
@@ -68,19 +113,24 @@ INSTALL_PATH="/opt/seagate/""${PRODUCT}"
 
 mkdir -p "$INSTALL_PATH"
 
+# build addb plugin
+if [ "$BUILD_ADDB" == "true" ]; then
+  echo "building addb plugin and adding this binary file as a part of cortx-rgw-integration rpm."
```
Comment (rahul-tripathi-git) [inline jenkins/build.sh:118]: instead if 'this binary file', should we name the file?

**Review by rahul-tripathi-git — APPROVED** [review]
Approved the PR. Couple of minor comments.

File: .gitignore [hunk 3]
```diff
@@ -4,3 +4,6 @@
 build/
 cortx_rgw_integration.egg-info/
 dist/
+src/addb_plugin/rgw_addb_map.o
+src/addb_plugin/rgw_addb_plugin.o
+src/addb_plugin/rgw_addb_plugin.so
```
Comment (rahul-tripathi-git) [inline .gitignore:9]: Why are these files put to gitignore list?

File: jenkins/build.sh (hunk 1 above)
Comment (sachitanands) [reply jenkins/build.sh]: done

File: jenkins/build.sh (hunk 2 above)
Comment (sachitanands) [reply jenkins/build.sh:128]: yes. added the name of binary file.

**Review by shriya-deshmukh — APPROVED** [review]

File: jenkins/build.sh [hunk 4]
```diff
@@ -68,19 +123,25 @@ INSTALL_PATH="/opt/seagate/""${PRODUCT}"
 
 mkdir -p "$INSTALL_PATH"
 
+# build addb plugin
+if [ "$BUILD_ADDB" == "true" ]; then
+  echo "building addb plugin and adding addb binary file (rgw_addb_plugin.so)\
```
Comment (cdeshmukh) [inline jenkins/build.sh:128]: echo "Building addb plugin(rgw_addb_plugin.so) for RGW and bundling same \
  in cortx-rgw-integration rpm."

File: .gitignore [hunk 5]
```diff
@@ -4,3 +4,6 @@
 build/
 cortx_rgw_integration.egg-info/
 dist/
+src/addb_plugin/rgw_addb_map.o
```
[the remark this answers is not in the document]
Comment (sachitanands) [reply .gitignore:7]: sure let me check on this.

File: jenkins/build.sh (hunk 4 above)
Comment (sachitanands) [reply jenkins/build.sh]: done

File: jenkins/build.sh [hunk 6]
```diff
@@ -18,39 +18,94 @@ set -e
...
 BUILD_NUMBER=
 GIT_VER=
 PRODUCT="cortx"
+ADDB_PLUGIN_DIR="$BASE_DIR"/src/addb_plugin
+MOTR_REPO="$BASE_DIR"/../cortx-motr
+RGW_REPO="$BASE_DIR"/../cortx-rgw
 
 usage() {
-    echo """usage: $PROG [-v version] [-g git_version] [-b build_number]""" 1>&2;
+    echo "usage: sh $PROG [--v version] [--g git_version] [--b build_number] [--build_addb] [--h]
```
[the remark this answers is not in the document]
Comment (sachitanands) [reply jenkins/build.sh]: done

**Comment by cdeshmukh** [comment]
address codacy warnings also

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (sachitanands) [reply .gitignore]: done. added cleanup function to delete those temporary files after building rgw rpm.

File: src/addb_plugin/README.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
+# Building addb plugin
+ Addb plugin (rgw_addb_plugin.so) is build as part of cortx-rgw-integration rpm.
+ All addb files which are required to build addb plugin are present as part of cortx-rgw-integration repo.
+
+## Dependencies for building addb plugin
+For building addb, we require two repositories to be cloned.
```
[the remark this answers is not in the document]
Comment (sachitanands) [reply src/addb_plugin/README.md]: done

File: src/addb_plugin/README.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  b) generate addb plugin using below commands,
+     cd <cortx-rgw-integration>/src/addb_plugin
+     make clean
+     make plugin
+
+## How addb plugin will be generatd in custom-ci/jenkins job
+For building addb plugin using custom-ci/jenkins job,
+  a) cortx-rgw-integration's build script will require additional parameter
+     e.g. sh build.sh --addb_build
+  b) this will be handled by RE team in jenkins's script.
```
[the remark this answers is not in the document]
Comment (sachitanands) [reply src/addb_plugin/README.md]: done

**Review by cdeshmukh — APPROVED** [review]

**Comment by sachitanands** [comment]
Triggered docker image job : 
https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/6337/  (In-Progress)

**Comment by sachitanands** [comment]
Deployment was successful and tested those changes with running IO ops.
image

**Pull request merged**